contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1051 @@
+     *   Held Token: (Held Token Balance) * (OP / Total OP)
+     *
+     * @param  buckets      The bucket numbers to withdraw from
+     * @param  maxWeights   The maximum weight to withdraw from each bucket. The amount of tokens
+     *                      withdrawn will be at least this amount, but not necessarily more.
+     *                      Withdrawing the same weight from different buckets does not necessarily
+     *                      return the same amounts from those buckets. In order to withdraw as many
+     *                      tokens as possible, use the maximum uint256.
+     * @param  beneficiary  The address to send the tokens to
+     * @return              1) The number of owedTokens withdrawn
+     *                      2) The number of heldTokens withdrawn
+     */
+    function withdraw(
+        uint256[] buckets,
+        uint256[] maxWeights,
+        address beneficiary
+    )
+        external
+        returns (uint256, uint256)
+    {
+        require(
+            beneficiary != address(0),
+            "BucketLender#withdraw: Beneficiary cannot be the zero address"
+        );
+        require(
+            buckets.length == maxWeights.length,
+            "BucketLender#withdraw: The lengths of the input arrays must match"
+        );
+
+        rebalanceBuckets();
+
+        uint256 totalOwedToken = 0;
+        uint256 totalHeldToken = 0;
+
+        uint256 maxHeldToken =
+            Margin(DYDX_MARGIN).isPositionClosed(POSITION_ID) ?
+            TokenInteract.balanceOf(HELD_TOKEN, address(this)) :
+            0;
+
+        for (uint256 i = 0; i < buckets.length; i++) {
+            (uint256 owedTokenForBucket, uint256 heldTokenForBucket) = withdrawInternal(
+                buckets[i],
+                maxWeights[i],
+                maxHeldToken
+            );
+
+            totalOwedToken = totalOwedToken.add(owedTokenForBucket);
+            totalHeldToken = totalHeldToken.add(heldTokenForBucket);
+        }
+
+        // Transfer share of owedToken
+        TokenInteract.transfer(OWED_TOKEN, beneficiary, totalOwedToken);
+        TokenInteract.transfer(HELD_TOKEN, beneficiary, totalHeldToken);
+
+        return (totalOwedToken, totalHeldToken);
+    }
+
+    // ============ Public State-Changing Functions ============
+
+    /**
+     * Allow anyone to refresh the bucket amounts if part of the position was closed since the last
+     * position increase.
+     */
+    function rebalanceBuckets()
+        public
+    {
+        if (Margin(DYDX_MARGIN).isPositionClosed(POSITION_ID)) {
+            return;
+        }
+
+        uint256 marginPrincipal = getCurrentPrincipalFromMargin();
+
+        accountForClose(principalTotal.sub(marginPrincipal));
+
+        assert(principalTotal == marginPrincipal);
+    }
+
+    /**
+     * Helper function to make sure allowance is set on the dYdX proxy. Callable by anyone.
+     */
+    function setMaximumAllowanceOnProxy()
+        public
```

Comment:
We could, but that also requires an external call and a state read every time. If we really are trying to enforce the same loanhash every time, then we shouldn't need to do this more than once (in the constructor) because the `loanFills` amount in the base protocol would overflow at the exact same time.

I really don't feel like it's a problem to set it to max int once. You would need a token with 18 decimal places and then a supply of like 10^80. I realize that the same tokens can be lent multiple times, but honestly I think you would run out of ether to lend that many tokens of anything.

Consider a gas cost of 1 gwei and that lending costs at least 1000 gas (a huge underestimate). Then a transaction costs 10^-6 eth. Since there is 10^8 eth in existence, then you could make 10^14 loans if you had all the eth in existence. Since 2**256 is around 10^77, then there would need to be at least 10^60 supply of a coin which is unheard of.

Reply:
I'd think it'd be more likely with a token with like 64 decimal places or something. The convention of 18 decimal places is just a convention. I know most tokens follow this, but could be a confusing edge case if it starts throwing errors for specific weird tokens

Also, does anything actually enforce the same loan hash each time